API: add dtype= option to python parser

chris-b1 opened the pull request:
- [x] part of #12686
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

Ultimately I'm working towards #8212 (types in excel parser), which should be pretty straightforward after this.

Right now the tests are moved from `c_parser_only.py`, may need to add some more
cc @gfyoung

chris-b1 committed:
API: add dtype= option to python parser

chris-b1 committed:
API: add dtype= option to python parser

chris-b1 committed:
remove unsupported test

chris-b1 committed:
remove unsupported test

chris-b1 committed:
add test/fix for dtype=object

chris-b1 committed:
add test/fix for dtype=object

chris-b1 committed:
float precision...

chris-b1 committed:
float precision...

chris-b1 committed:
float precision fix

chris-b1 committed:
float precision fix

jorisvandenbossche commented on pandas/io/parsers.py:
```diff
@@ -1284,48 +1285,83 @@ def _agg_index(self, index, try_parse_dates=True):
         return result
 
-    def _convert_types(self, values, na_values, try_num_bool=True):
+    def _infer_types(self, values, na_values, try_num_bool=True):
```
While you are at it, can you add a docstring here?

jorisvandenbossche commented on pandas/io/parsers.py:
```diff
@@ -110,8 +114,9 @@
-    not interpret dtype.
+    Use `str` or `object` to preserve and not interpret dtype.
+    If converters are specified, they will be applied AFTER
+    dtype conversion.
```
Is there a test that exercises this assumption?

chris-b1 replied:
I was wrong, the c-parser actually ignores the dtype and just uses the converter - changed docstring and added matching test.

gfyoung commented on pandas/io/parsers.py:
```diff
@@ -1284,48 +1285,83 @@ def _agg_index(self, index, try_parse_dates=True):
+                    values, conv_f, na_values,
+                    col_na_values, col_na_fvalues)
+            else:
+                try_num_bool = True
```
Why not just this:
~~~python
try_num_bool = cast_type and is_object_dtype(cast_type)
~~~

gfyoung replied:
Why not just this:
~~~python
try_num_bool = cast_type and is_object_dtype(cast_type)
~~~

jorisvandenbossche reviewed (commented):
Looks nice to me! (added two minor comments)

gfyoung commented on pandas/io/parsers.py:
```diff
@@ -1355,6 +1392,23 @@ def _convert_types(self, values, na_values, try_num_bool=True):
         return result, na_count
 
+    def _cast_types(self, values, cast_type, column):
+        """ cast column to type specified in dtypes= param """
```
While I know this isn't done too much for internal functions, it would be good to start documenting these functions more thoroughly for development purposes.

chris-b1 committed:
add docs; test for conv cast

chris-b1 committed:
add docs; test for conv cast

jorisvandenbossche commented:
@chris-b1 xref #14558. You say that `dtype` is ignored when `converters` is specified. In that case it would be nice to raise a warning about this if both are specified I think.

chris-b1 committed:
Add warning if both converter and dtype specified

chris-b1 committed:
Add warning if both converter and dtype specified

jorisvandenbossche commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -31,7 +31,15 @@ Other enhancements
 
+.. ipython:: python
+
+   from io import StringIO
```
this will not work on python 2. Should check how it is done in other places in the docs

jorisvandenbossche replied:
`from pandas.compat import StringIO` is imported in a suppressed ipython code block at the top of the file, so you can just use `StringIO`

chris-b1 replied:
For the record, the `io` module was backported to 2.6, but I will remove this, as it is already imported like you mentioned.

jorisvandenbossche commented on pandas/io/parsers.py:
```diff
@@ -110,8 +114,12 @@
+    If converters are specified, they will be applied INSTEAD
+    of dtype conversion.
+
+  .. versionadded:: 0.20.0 support for the Python parser.
```
I personally would leave this out of the docstring (this is already so long ..), but not strong feelings, your take

jorisvandenbossche reviewed (commented):
Two small comments, looks good to me!

chris-b1 committed:
doc comments

chris-b1 committed:
doc comments

chris-b1 commented:
I think I've got all the feedback so far worked in - any more comments?  @jorisvandenbossche, @gfyoung, @jreback

gfyoung reviewed (approved):
LGTM!

jorisvandenbossche reviewed (approved):
Looks good to me as well

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -32,6 +32,14 @@ Other enhancements
 
 - ``pd.read_excel`` now preserves sheet order when using ``sheetname=None`` (:issue:`9930`)
 
```
i would make a sub-section

jreback commented on pandas/parser.pyx:
```diff
@@ -1038,14 +1038,33 @@ cdef class TextReader:
+
             if conv:
+                if col_dtype is not None:
+                    warnings.warn(("Both a converter and dtype were specified "
```
is there a test for this? (IOW, this was prob tested for c-engine, but now need to test for all)

jorisvandenbossche replied:
yes, this was added here:https://github.com/pandas-dev/pandas/pull/14295/commits/e0d16067cb5aded07ba92de9eb61478f07e66451

gfyoung commented:
@chris-b1 : I'm putting together a PR to patch #14712.  As this directly relates to your PR, I would suggest waiting for mine to make sure the bug isn't an issue for you.

jorisvandenbossche commented:
@chris-b1 can you rebase this? Then this can be merged I think (depending on the status of @gfyoung new PR)

chris-b1 committed:
doc updates

chris-b1 committed:
doc updates

chris-b1 commented:
Finally got back to this, rebased and updated docs - assuming tests still pass this should be good.  Will have to check, but I don't think this will conflict with #14717.

chris-b1 committed:
lint

chris-b1 committed:
lint

jorisvandenbossche committed:
TST: move empty dtype tests from c_parser_only to dtype tests

jorisvandenbossche commented:
@chris-b1 I merged #14717. This added a dtype test to `c_parser_only`, so added a commit to move this to the general dtype tests as well. Feel free to merge if you're ok with this and the tests are green.

chris-b1 committed:
Merge branch 'textreader-dtype' of https://github.com/chris-b1/pandas into textreader-dtype

Conflicts:
	pandas/io/tests/parser/dtypes.py

chris-b1 committed:
issue ref

jorisvandenbossche committed:
fix merge conflict leftover

jorisvandenbossche commented:
@chris-b1 Thanks!
